### ISSUE TITLE
disable-tls1 & disable-tls1_1 don't work.  TLS 1.0 and 1.1 still usable when Configured to be disabled.

### ISSUE THREAD

**Issue by JacobBrownAustin** [openssl/openssl#7048]
If you Configure your own Openssl build to have TLS 1.0 and/or TLS 1.1 disabled, it is still possible for them to be used.

This can cause issues if you want to do things such as make sure you are following the PCI DSS 3.2.1 , or otherwise just care a lot about security and don't want to have older TLS versions enabled.

There are two main use cases I've found that have bugs with this.

A)  Your client specifies they want to connect to a specific version of TLS in the TLS options.

B) Your client uses a "method" for a specific version of TLS.

In both of those cases, you would think that it must not be possible to do if you have configured your openssl libraries to be built with those versions disabled.

I have made a functional test that shows this bug.  https://github.com/JacobBrownAustin/jacob-misc/blob/master/OpenSSL/Only-TLS-1.2/OpenSslTlsTest-1.2-only-ssllabs.php
In order to run this test, you must
A) have your openssl configured to have TLS 1.0 and TLS 1.1 disabled (as long as SSL 2 and SSL 3)
B) be able to connect to https://www.ssllabs.com of the different ports.
C) Have php installed and linked against libssl (openssl extension enabled) and curl extension enabled (also linked against libssl)
D) Have phpunit installed

I have made proof of concept Debian packages that have these bugs fixed (and compiled with TLS 1.0 and 1.1 disabled).
You can find them here:
1.0_1.0.2l:
https://github.com/JacobBrownAustin/jacob-misc/tree/master/OpenSSL/Only-TLS-1.2/Debian-Packages/Stretch/openssl1.0_1.0.2l%2Bjacob2
1.1.0f:
https://github.com/JacobBrownAustin/jacob-misc/tree/master/OpenSSL/Only-TLS-1.2/Debian-Packages/Stretch/openssl_1.1.0f-3%2Bjacob1

**Comment by JacobBrownAustin** [comment]
Created pull requests to fix this issues in 1.0.2 and 1.1.0.

**Comment by mattcaswell** [comment]
> A) Your client specifies they want to connect to a specific version of TLS in the TLS options.

> B) Your client uses a "method" for a specific version of TLS.

This is a deliberate and documented behaviour for those options. The intent of them is to disable the automatic negotiation of those protocol versions.  If you want to additionally disable the explicit "method" versions you should use disable-tls1_1-method and disable-tls1-method.

https://github.com/openssl/openssl/blob/837017b4748d587912d9d218894644d6ca86721f/INSTALL#L469-L482

**Comment by JacobBrownAustin** [comment]
@mattcaswell That didn't work (when I tried it a few months ago at least) because when you compile it with those options, it breaks any of the programs linked against libssl.so because of missing symbols.  The pull requests I made disables the functionality without breaking any programs that are dependent on libssl.so .  That way, a system administrator can change which protocols are disabled without having to recompile all the programs that use openssl libraries (which would be very burdensome, and the cases of closed source software, impossible).

**Comment by mattcaswell** [comment]
> That didn't work (when I tried it a few months ago at least) because when you compile it with those options, it breaks any of the programs linked against libssl.so because of missing symbols. 

Which is why we have separated these two things into two different options. It sounds like what you are wanting is yet a third option - which leaves the symbols there but (presumably) leaves them as inoperable? We cannot modify the fundamental behaviour of the existing options (at least not until we do some future 1.2.0 version) because that would be a breaking change.

I can't say I'm particularly enthusiastic to add a third way of doing this. I'd be interested in @kroeckx view on this.

**Comment by JacobBrownAustin** [comment]
Well, I don't care what the solution to this is.  I was just looking for a way to disable older TLS versions (and SSL) with only having to recompile OpenSSL, but not have to recompile any programs that use OpenSSL libraries.

(I had assumed that disable-tls1 was supposed to do this, but if not, it might want to be better documented for people who are looking to do things like PCI DSS compliance.)

So, it doesn't matter to me if the way to do this is a third compile option, or to change the functionality of `disable-tls1-method` so that it still compiles the methods, but changes the methods so that they just throw unsupported version errors, or to have a solution like my pull request that modifies the behaviour of the `disable-tls1` option.

**Comment by mattcaswell** [comment]
I could see changing the meaning of disable-tls1-method as workable. That wouldn't actually break anything.

**Comment by JacobBrownAustin** [comment]
Okay I'll try that later today or maybe this weekend and see if I can get it working.

**Comment by mattcaswell** [comment]
I'd still like to hear the opinion of @kroeckx before you go too far with it.

**Comment by kroeckx** [comment]
At least for SSLv3 we have enable-ssl3 and enable-ssl-method. We
had the same for SSLv2. We could do this same split for other
protocols.

**Comment by kroeckx** [comment]
I guess I might have misunderstood something.

**Comment by kroeckx** [comment]
It make sense for me to have no-tls1 disable all tls1 functionality, and make the TLS1 methods return NULL, and with no-tls1-method also remove the function itself. This at least looks more what you expect when you don't read the documentation.

Note that we need those function internally, so if you want SSL3 enabled and use the SSLv23_method/TLS_method() you actually need to use enable-ssl3 enable-ssl3-method, so that we can internally make use of the method.

So I think we currently have:

option | Use in TLS_method | TLS1_method
-- | - | -
--  | yes | works
no-tls1 | no | works
no-tls1-method | no | not available
no-tls1 no-tls1-method | no | not available

While we might want:

option | Use in TLS_method | Use TLS1_method
-- | - | -
-- | yes | works
no-tls1 | no | returns NULL
no-tls1-method | yes | not available
no-tls1 no-tls1-method | no | not available

**Comment by kroeckx** [comment]
@vdukhovni: Do you have any comment on this?

**Comment by vdukhovni** [comment]
> @kroeckx @vdukhovni: Do you have any comment on this?

I am not opposed to retaining stub definitions of `TLS1_method()`, `TLS1_client_method()` and `TLS1_server_method()`, ... that return NULL when the method is disabled.  Debian did this for a while with SSLv2, and I don' recall any adverse effects.

Applications that try to use these functions, presumably with `SSL_CTX_new(TLS1_method())` or similar will get an error, as intended, but will compile and work provided they avoid trying to use the disabled methods.

**Comment by mattcaswell** [comment]
> It make sense for me to have no-tls1 disable all tls1 functionality

If we weren't starting from here then I'd agree. But I don't think we can do that now - that would be a breaking change.

**Comment by vdukhovni** [comment]
> @mattcaswell If we weren't starting from here then I'd agree. But I don't think we can do that now - that would be a breaking change.

Definitely agreed. The present semantics of "no-tls1" should remain.  What the request is asking for is much more akin to "no-tls1-method", **but** with the method stubbed out to return NULL rather than removed.  We can probably accomodate that, in the interests of ABI compatibility for the resulting build.

**Comment by nhorman** [comment]
Marking as inactive, to be closed at the end of 3.4 dev barring further input

**Comment by t8m** [comment]
Closing in favor of #28343
